pkg/object/mqttproxy/topic.go

```diff
@@ -0,0 +1,332 @@
+		levelMgr: newTopicLevelManager(),
+	}
+}
+
+func newTopicLevelManager() *topicLevelManager {
+	return &topicLevelManager{
+		data: make(map[string]*topicLevels),
+	}
+}
+
+func validTopic(levels []string) bool {
+	for i, level := range levels {
+		if strings.ContainsRune(level, '+') && len(level) != 1 {
+			return false
+		}
+		if strings.ContainsRune(level, '#') {
+			if len(level) != 1 || i != len(levels)-1 {
+				return false
+			}
+		}
+	}
+	return len(levels) > 0
+}
+
+func (mgr *TopicManager) getLevels(topic string) ([]string, error) {
+	return mgr.levelMgr.get(topic)
+}
+
+func (mgr *TopicManager) subscribe(topics []string, qoss []byte, clientID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	for i, t := range topics {
+		if err := mgr.insert(t, qoss[i], clientID); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (mgr *TopicManager) unsubscribe(topics []string, clientID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	for _, t := range topics {
+		if err := mgr.remove(t, clientID); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (mgr *TopicManager) findSubscribers(topic string) (map[string]byte, error) {
+	mgr.RLock()
+	defer mgr.RUnlock()
+	return mgr.match(topic)
+}
+
+func (mgr *TopicManager) insert(topic string, qos byte, clientID string) error {
+	levels, err := mgr.getLevels(topic)
+	if err != nil {
+		return err
+	}
+	node := mgr.root
+	var nextNode *topicNode
+	var ok bool
+	for _, l := range levels {
+		nextNode, ok = node.nodes[l]
+		if !ok {
+			nextNode = newNode()
+			node.nodes[l] = nextNode
+		}
+		node = nextNode
+	}
+	node.clients[clientID] = qos
+	return nil
+}
+
+func (mgr *TopicManager) remove(topic string, clientID string) error {
+	levels, err := mgr.getLevels(topic)
+	if err != nil {
+		return err
+	}
+	node := mgr.root
+	var nextNode *topicNode
+	var ok bool
+	prevNodes := []*topicNode{}
+	for _, l := range levels {
+		if nextNode, ok = node.nodes[l]; !ok {
+			return nil
+		}
+		prevNodes = append(prevNodes, node)
+		node = nextNode
+	}
+	delete(node.clients, clientID)
+
+	// clear memory
+	for i := len(prevNodes) - 1; i >= 0; i-- {
+		node = prevNodes[i].nodes[levels[i]]
+		if len(node.clients) == 0 && len(node.nodes) == 0 {
+			delete(prevNodes[i].nodes, levels[i])
+		} else {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (mgr *TopicManager) match(topic string) (map[string]byte, error) {
+	levels, err := mgr.getLevels(topic)
+	if err != nil {
+		return nil, err
+	}
+	ans := make(map[string]byte)
+
+	currentLevelNodes := []*topicNode{mgr.root}
+	nextLevelNodes := []*topicNode{}
+	for _, topicLevel := range levels {
+		for _, node := range currentLevelNodes {
+			for nodeLevel, nextNode := range node.nodes {
+				if nodeLevel == "#" {
+					nextNode.addClients(ans)
+				}
+				if nodeLevel == "+" || nodeLevel == topicLevel {
+					nextLevelNodes = append(nextLevelNodes, nextNode)
+				}
+			}
+		}
+		currentLevelNodes = nextLevelNodes
+		nextLevelNodes = []*topicNode{}
```

Comment:
As an initialization, you can remove code line#167 and line#180, and put it after line#168.